### DISCUSSION TITLE
How to add Gameobjects at runtime

### DISCUSSION THREAD

**Discussion by Bhuelsmann** [bijington/orbit#31] [Q&A] [answered]
Hi,

I am playing a little bit around with your engine. Nice work so far! :)
Now I am trying to add some new Gameobjects to a scene, while the game is running.
I tried it, with adding a method to the scene which adds new gameobjects to his container, but it didn't work.
Do you have any hints for me?

With this clickevent I try to add a new object. Balloon is the gameobject, BalloonScene is my Scene.

` private void OnAddBalloonClicked(object sender, EventArgs e)
{  
try
{
var balloon = new Balloon(gameSceneManager); 
balloon.SetBalloonData(new BalloonData() { Speed = 3 }); 
balloonScene.AddBalloon(balloon); 
}
catch (Exception ex)
 {
 throw ex;
}
 }`

In the scene is just:

`        public void AddBalloon(Balloon balloon)
        {
            Add(balloon);
        }`

In debug I can see, that the objects are added to the collection, but they are not rendered. You cannot see them on screen. They have another position , so its not, that they are stacked in z-axis or something else.

Thx for your help :)

**Comment by bijington** [comment]
Thank you :).

In theory that should work. As I am doing something similar in this file around line 84: https://github.com/bijington/orbit/blob/main/games/Orbit/Orbit/GameObjects/AsteroidLauncher.cs

Is your source open so I could take a look? If not can you confirm that you have called Start from the game scene manager so that the rendering loop is active?

**Reply by bijington** [reply]
It's also worth noting that while you can set the `Bounds` property at the moment you are still responsible for using that in your Render method. I do plan to encapsulate the rendering at some point but I wanted to try out many different games before finding an API surface that I like.

**Reply by Bhuelsmann** [reply]
Hi, 
I don't have the code public. At the moment it is only some testing code, one scene and on gameobject. 
Yes the scene is started. My first gameobject is added in the scene constructor and I can see it "flying" from right to left.

`        public BalloonScene(Balloon balloon)
        {
            Add(balloon);
        }`
But when I call the AddBalloon method, nested in the sceneclass, it added new objects in the list, but didn't render it.
I added some consoleoutput in the gameobjects to see the positions, but there is only the position from the initial object.

It feels like the new objects are not activated or the scene didn't recognize them.

**Comment by Bhuelsmann** [comment]
I tried a little bit around, but couldn't solve it.
In next step I tried to "copy" your AstroidLuncher. So I created my BalloonSender:

`        

        public BalloonSender(
        IServiceProvider serviceProvider,
        IGameSceneManager gameSceneManager)
        {
            this.serviceProvider = serviceProvider;
            this.gameSceneManager = gameSceneManager;
        }

        public override void Update(double millisecondsSinceLastUpdate)
        {
            if ((lastUpdate - lastSpawn).TotalSeconds > 2)
            {
                var balloon = new Balloon(gameSceneManager);
                var balloonData = new BalloonData() { Speed = 3, YPosition = (new Random()).Next(500) };
                balloon.SetBalloonData(balloonData);
                
                CurrentScene.Add(balloon);

                lastSpawn = DateTime.UtcNow;
            }

            lastUpdate = DateTime.UtcNow;
        }`

I added it to the ServiceCollection and in the scene I added it as gameobject. But now my app crashes with an 
"Value does not fall within the expected range." exception.
It crashes in the GameSceneManager at line 113

`        dispatcher.DispatchDelayed(
            TimeSpan.FromMilliseconds(delayUntilNextUpdate),
            () =>
            {
                UpdateScene();
            });`
Any hint what I am doing wrong?

**Comment by Bhuelsmann** [comment]
Hi, 
I added my project as attachment, would be great if you find time to have a short look on it.

[TestProject Orbit Engine.zip](https://github.com/bijington/orbit/files/13522392/TestProject.Orbit.Engine.zip)

greetings

**Reply by bijington** [reply]
Thanks for sharing. I can try to take a look over the weekend

**Reply by bijington** [reply, marked as answer]
I have had a quick look at the sample and I think I know what the issue is, you are registering `BalloonScene` as Transient and will ultimately end up with 2 instances of `BalloonScene`:
- One in the backing field `balloonScene` in your `MainPage` class
- One is also created when you call `gameSceneManager.LoadScene<BalloonScene>(GameView);`

I would recommend multiple options to solve this issue:
1. Modify the registration of `BalloonScene` to use `AddSingleton` or even `AddScoped` 
e.g. 

```csharp
services
            .AddScoped<BalloonScene>();
```

2. Modify your `OnAddBalloonClicked` method to access the instance of `BalloonScene` loaded into the `gameSceneManager`. You could modify your code to look something like:

```csharp
private void OnAddBalloonClicked(object sender, EventArgs e)
{
    var balloon = new Balloon(gameSceneManager);
    var balloonData = new BalloonData() { Speed = 3, YPosition = (new Random()).Next(500) };
    balloon.SetBalloonData(balloonData);
    ((BalloonScene)gameSceneManager.CurrentScene).AddBalloon(balloon);
}
```

To save casting each time the button is clicked you could alternatively change your constructor to store the loaded scene into your backing field, e.g.

```csharp
public MainPage(IServiceProvider serviceProvider, IGameSceneManager gameSceneManager)
{
    InitializeComponent();
    this.serviceProvider = serviceProvider;
    this.gameSceneManager = gameSceneManager;
    gameSceneManager.LoadScene<BalloonScene>(GameView);

    this.balloonScene = (BalloonScene)gameSceneManager.CurrentScene;
}
```

**Reply by Bhuelsmann** [reply]
Thx, for your help. Its working, now :)

**Reply by bijington** [reply]
That's great to hear! Thank you for using this library 🎮

**Reply by bijington** [reply]
And let me know once the game is playable, I'll be a keen tester 😄
